File: dataStructures/heaps/findKthLargestHeap.js
// Leetcode: 215. Kth Largest Element in an Array
// URL: https://leetcode.com/problems/kth-largest-element-in-an-array/
// Time: O(n) average case, O(n^2) worst case
// Space: O(1)

// Given an integer array nums and an integer k, return the kth largest element in the array.
// Note that it is the kth largest element in the sorted order, not the kth distinct element.
// Can you solve it without sorting?

// Explanation:
// The algorithm efficiently finds the kth largest number by dividing the list into smaller parts and focusing only on the part that contains the kth largest number. This way, it avoids sorting the entire list, making it faster on average.

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function (nums, k) {
  // get complementK (for the for loop later on)
  let complementK = nums.length - k;
  // create minHeap class with the following properties: parentIndex, leftChildIndex, rightChildIndex, swap, insert, removeMin, heapifyUp, heapifyDown
  class MinHeap {
    constructor() {
      this.heap = [];
    }
    buildHeap(arr) {
      for (let i = 0; i < arr.length; i++) {
        this.insert(arr[i]);
      }
    }
    parentIndex(index) {
      return Math.floor((index - 1) / 2);
    }
    leftChildIndex(index) {
      return 2 * index + 1;
    }
    rightChildIndex(index) {
      return 2 * index + 2;
    }
    swap(i, j) {
      [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
    }
    insert(val) {
      this.heap.push(val);
      this.heapifyUp(this.heap.length - 1);
    }
    removeMin() {
      if (this.heap.length === 0) return null;
      if (this.heap.length === 1) return this.heap.pop();
      const min = this.heap[0];
      this.heap[0] = this.heap.pop();
      this.heapifyDown(0);
      return min;
    }
    heapifyUp(index) {
      let currIndex = index;
      while (
        currIndex > 0 &&
        this.heap[currIndex] < this.heap[this.parentIndex(currIndex)]
      ) {
        this.swap(currIndex, this.parentIndex(currIndex));
        currIndex = this.parentIndex(currIndex);
      }
    }
    heapifyDown(index) {
      // make currSmallestIndex = index
      let currSmallestIndex = index;
      // get left and right child indexes
      let leftChildIndex = this.leftChildIndex(index);
      let rightChildIndex = this.rightChildIndex(index);
      // if statements for each to determine if the child is smaller, and if so, then make currSmallestIndex equal to that childindex
      if (
        leftChildIndex < this.heap.length &&
        this.heap[currSmallestIndex] > this.heap[leftChildIndex]
      ) {
        currSmallestIndex = leftChildIndex;
      }
      if (
        rightChildIndex < this.heap.length &&
        this.heap[currSmallestIndex] > this.heap[rightChildIndex]
      ) {
        currSmallestIndex = rightChildIndex;
      }
      if (currSmallestIndex !== index) {
        // if currSmallestIndex does not equal index, then swap the two items
        this.swap(currSmallestIndex, index);
        // then call heapifyDown recursively on that index
        this.heapifyDown(currSmallestIndex);
      }
    }
  }

  // create a new MinHeap of nums (automatically sorted)
  let minHeap = new MinHeap();
  minHeap.buildHeap(nums);

  // for complementK times, remove min,
  for (let i = 0; i < complementK; i++) {
    minHeap.removeMin();
  }

  // then remove min once more and return it
  return minHeap.removeMin();
};

File: dataStructures/heaps/findKthLargestHeap.py
import heapq
import math
from typing import List

# this is my long solution, an adaptation of the js solution, but in python you actually don't need to create a minHeap class. You can just use the heapq module, which is demonstrated below.


class Solution:
    # create minHeap class with the following: parentIndex, leftChildIndex, rightChildIndex, swap, insert, removeMin, heapifyUp, heapifyDown
    class MinHeap:
        def __init__(self):
            self.heap = []

        def build_heap(self, arr):
            for i in range(len(arr)):
                self.insert(arr[i])

        def parent_index(self, index):
            return math.floor((index - 1) / 2)

        def left_child_index(self, index):
            return 2 * index + 1

        def right_child_index(self, index):
            return 2 * index + 2

        def swap(self, i, j):
            self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

        def insert(self, val):
            self.heap.append(val)
            self.heapify_up(len(self.heap) - 1)

        def heapify_up(self, index):
            curr_index = index
            while curr_index > 0 and self.heap[curr_index] < self.heap[self.parent_index(curr_index)]:
                self.swap(self.parent_index(curr_index), curr_index)
                curr_index = self.parent_index(curr_index)

        def remove_min(self):
            if len(self.heap) == 0:
                return None
            if len(self.heap) == 1:
                return self.heap.pop(0)
            min_val = self.heap[0]
            self.heap[0] = self.heap.pop(-1)
            self.heapify_down(0)
            return min_val

        def heapify_down(self, index):
            curr_smallest_index = index
            left = self.left_child_index(index)
            right = self.right_child_index(index)

            if left < len(self.heap) and self.heap[left] < self.heap[curr_smallest_index]:
                curr_smallest_index = left

            if right < len(self.heap) and self.heap[right] < self.heap[curr_smallest_index]:
                curr_smallest_index = right

            if curr_smallest_index != index:
                self.swap(curr_smallest_index, index)
                self.heapify_down(curr_smallest_index)

    def findKthLargest(self, nums: List[int], k: int) -> int:
        complement_k = len(nums) - k
        min_heap = self.MinHeap()
        # Build the min heap with all elements
        min_heap.build_heap(nums)
        # Remove elements until the heap size is complement_k
        for _ in range(complement_k):
            min_heap.remove_min()
        # The next minimum is the kth largest
        result = min_heap.remove_min()
        return result


# this is the short solution using the heapq module
# the heapq module is a min heap, so we can just push the first k elements into the heap, and then for the remaining elements, we can push the element

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # Initialize a min heap with the first k elements
        min_heap = nums[:k]
        heapq.heapify(min_heap)

        # Iterate through the remaining elements
        for num in nums[k:]:
            if num > min_heap[0]:  # If current element is larger than the smallest in heap
                # Push new element and pop the smallest
                heapq.heappushpop(min_heap, num)

        # The root of the heap is the kth largest element
        return min_heap[0]
